tinywl/tinywl.c

```diff
@@ -125,8 +124,13 @@ static void focus_view(struct tinywl_view *view, struct wlr_surface *surface) {
 	 * track of this and automatically send key events to the appropriate
 	 * clients without additional work on your part.
 	 */
-	wlr_seat_keyboard_notify_enter(seat, view->xdg_surface->surface,
-		keyboard->keycodes, keyboard->num_keycodes, &keyboard->modifiers);
+	struct wlr_keyboard *keyboard = wlr_seat_get_keyboard(seat);
+	if (keyboard != NULL) {
+		wlr_seat_keyboard_notify_enter(seat, view->xdg_surface->surface,
+			keyboard->keycodes, keyboard->num_keycodes, &keyboard->modifiers);
+	} else {
+		wlr_seat_keyboard_notify_enter(seat, surface, NULL, 0, NULL);
+	}
```

Comment:
This kinda makes me wonder while wlroots makes compositors do this check themselves, instead `wlr_seat_keyboard_notify_enter()` could just take a pointer to a `wlr_keyboard`.

Reply:
Hm, yes, that's a good point. At some point we wanted to get rid of `wlr_seat_set_keyboard` and de-couple `wlr_keyboard` from `wlr_seat` (just like `wlr_cursor`/`wlr_pointer` is decoupled from `wlr_seat`), but stopped halfway through. Ref https://github.com/swaywm/wlroots/issues/803